File: book/surfaces_demo.c
/* SPDX-License-Identifier: Apache-2.0 */
/* Copyright 2026 lituus-lab */
/* Run by book/surfaces.nim during the book build; its output is the page's. */
#include <stdio.h>
#include "UniChecksum.h"

int main(void) {
  const uint8_t check[] = {'1','2','3','4','5','6','7','8','9'};
  uint32_t state;

  printf("unichecksum_version()             = %s\n", unichecksum_version());
  printf("unichecksum_crc32(\"123456789\")    = %08X\n",
         unichecksum_crc32(check, sizeof check));
  printf("unichecksum_crc64(\"123456789\")    = %016llX\n",
         (unsigned long long)unichecksum_crc64(check, sizeof check));

  /* Out of domain: the boundary clamps rather than unwinding into C. */
  printf("unichecksum_crc32(NULL, 8)        = %08X   (nothing was read)\n",
         unichecksum_crc32(NULL, 8));
  state = unichecksum_crc32_init();
  printf("crc32_update(state, NULL, 8)      = %08X   (state, unchanged)\n",
         unichecksum_crc32_update(state, NULL, 8));
  return 0;
}

File: book/surfaces_demo.py
# SPDX-License-Identifier: Apache-2.0
# Copyright 2026 lituus-lab
# Run by book/surfaces.nim during the book build; its output is the page's.
# A file rather than `python3 -c`: the snippet needs both quote characters, and
# nesting them through Nim and then the shell is how a block ends up green and
# wrong.
import unichecksum as u

print("crc32(b'123456789')    =", hex(u.crc32(b"123456789")))
print("adler32(b'123456789')  =", hex(u.adler32(b"123456789")))
print("crc32(b'6789', crc32(b'12345')) =",
      hex(u.crc32(b"6789", u.crc32(b"12345"))))
print("crc32(b'123456789')             =", hex(u.crc32(b"123456789")))

for bad in ("a str", None):
    try:
        u.crc32(bad)
    except (TypeError, ValueError) as exc:
        print(f"  crc32({bad!r}) -> {type(exc).__name__}: {exc}")
try:
    u.crc32(b"x", -1)
except (TypeError, ValueError) as exc:
    print(f"  crc32(b'x', -1) -> {type(exc).__name__}: {exc}")
